flang/runtime/terminator.cpp

```diff
@@ -12,26 +12,47 @@
 
 namespace Fortran::runtime {
 
-[[noreturn]] void Terminator::Crash(const char *message, ...) const {
-  va_list ap;
-  va_start(ap, message);
-  CrashArgs(message, ap);
-  va_end(ap);
-}
-
-static void (*crashHandler)(const char *, int, const char *, va_list &){
+#if !defined(RT_DEVICE_COMPILATION)
+[[maybe_unused]] static void (*crashHandler)(const char *, int, const char *, va_list &){
     nullptr};
 
 void Terminator::RegisterCrashHandler(
     void (*handler)(const char *, int, const char *, va_list &)) {
   crashHandler = handler;
 }
 
-[[noreturn]] void Terminator::CrashArgs(
-    const char *message, va_list &ap) const {
+void Terminator::InvokeCrashHandler(const char *message, ...) const {
   if (crashHandler) {
+    va_list ap;
+    va_start(ap, message);
     crashHandler(sourceFileName_, sourceLine_, message, ap);
+    va_end(ap);
+  }
+}
+
+[[noreturn]] void Terminator::CrashArgs(
+    const char *message, va_list &ap) const {
+  CrashHeader();
+  std::vfprintf(stderr, message, ap);
+  va_end(ap);
+  CrashFooter();
+}
+#endif
+
+RT_OFFLOAD_API_GROUP_BEGIN
+
+RT_API_ATTRS void Terminator::CrashHeader() const {
+#if defined(RT_DEVICE_COMPILATION)
+  std::printf("\nfatal Fortran runtime error");
```

Comment:
`fprintf(stderr, ...)` here and below.

Reply:
Device compilers (at least some of them) do not support `stderr`.  In general, only a plain `printf` is supported on the device, e.g. https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#formatted-output

Reply:
Yes, I know, but `stderr` is being used in `Terminator::CrashArgs()` above, so I was hopeful that the limitation had been addressed.  But now your host `Terminator::CrashArgs()` calls a `CrashHeader()` and `CrashFooter()` that write to stdout while still printing the arguments via `vfprintf(stderr, ...)`.  And I don't see an implementation of `Terminator::CrashArgs()` for the device, but I might just be confused by the diff here.

Reply:
I had to comment out `Terminator::CrashArgs` for the device compilation (with `#if !defined(RT_DEVICE_COMPILATION)` at line 15), since neither `stderr` nor `va_list` are supported.  There are direct uses of `CrashArgs` in not yet enabled parts of the library, and I will have to deal with them as well.

Note that the `crashHandler` support is also not enabled for the device compilation.  It seems to be used only for in-tree testing, so I thought it would not be a problem (at least for now).

Reply:
Am I misreading the new code?  `CrashArgs()` calls `CrashHeader()` & `CrashFooter()` and they write to stdout.  But the `vfprintf()` writes to stderr.  Is that not a problem?

Reply:
`CrashHeader` and `CrashFooter` have different implementations for host and device.  On the device they only use `std::printf`, and on the host they use all the other functions to write to `stderr`.  `CrashArgs` does call them, but this only happens on the host, so everything is printed to `stderr`.

The code is different under `defined(RT_DEVICE_COMPILATION)` macro check.